site/docs/v1/tech/events-webhooks/signing.adoc

```diff
@@ -0,0 +1,164 @@
+
+Keys are generated or imported from [breadcrumb]#Settings -> Key Master#.  Webhooks can be signed with three types of keys
+
+- EC key - strongest cryptography, public key can be available
+- RSA key - strong cryptography, public key can be available
+- HMAC key - fast cryptography, requires manual key distribution
+
+EC and RSA keys allow you to make public keys available through the `/.well-known/jwks.json` endpoint, which facilitates key rotation.  If your webhook listener cannot make outbound network connections or you prefer to manually configure your key in your webhook listener, HMAC keys are a good option.
+
+For this example, we'll generate an RSA public key pair.
+
+Next, you configure your webhook to sign the event with this key.  From [breadcrumb]#Settings -> Webhooks#, click on the Edit button for your webhook (or create a new webhook).  Select the Signature tab panel.  Once you select the check box to enable webhook signatures, a select box allows you to choose the generated key.
+
+image::webhook-settings-signature.png[Webhook Settings Signature,width=1200,role=shadowed]
+
+=== Signature Verification
+
+The webhook signature is provided in the HTTP header `X-FusionAuth-Signature-JWT` as a signed JWT with a claim of `request_body_sha256` containing the sha256 hash of the webhook event payload.
+
+Your webhook listener can verify the signature by
+
+- Verify the JWT is properly signed
+- Decode the JWT
+- Compare the JWT's `request_body_sha256` claim against your own calculated sha256 hash of the event body
+
+[source,ini]
+.Example webhook HTTP header
+----
+X-FusionAuth-Signature-JWT: eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6Il9IMDd3VkcxZlYzbDVpaDc0ck54SUMzbmV2RSJ9.eyJyZXF1ZXN0X2JvZHlfc2hhMjU2IjoiS2VWKy9IR29JUXJ4dUU1WVBDUlI2QXVRT0p2ZWxkWU5OaGJWaTFpMjJxaz0ifQ.J70gqZVuTej8FfriQqJJZecCT6XOZKH6h6Te2ir_yrSwR3luhoj_R1vAZULdrktaFPqXFXbnq9prN8j3ddelUVA5SU51J-MWVhz1bkimLo8EEdJ47ytI_97rPqVK1YJ6FSiS8_o37gablaQZv2WDbZ6ap-t4hNU5m7uwZTW9DerKg9iQjMDUIlfafEwsROLfNPfK49IsCzBNCQ8SsinVbGU0dNbs9YfMAxNzSuEKdZOIXkRNgjPfWpPnkwBbroWUrrpcoAcBSQIYFajKV-MFRISnFZ_blYps16f95iQsuTfqBkBH3r59R5tFBP66FA1bvQJZVlAHJfdNTXnXx2F2BQ
+----
+
+The JWT decodes with:
+[source,json]
+.JWT header
+----
+{
+   "alg": "RS256",
+   "typ": "JWT",
+   "kid": "_H07wVG1fV3l5ih74rNxIC3nevE"
+ }
+----
+
+[source,json]
+.JWT payload
+----
+{
+  "request_body_sha256": "KeV+/HGoIQrxuE5YPCRR6AuQOJveldYNNhbVi1i22qk="
+}
+----
+
+The `kid` identifies the Id of the key used to sign the JWT.  JWT libraries can look the key up from the JWKS endpoint, or a locally stored key can be used.  After verifying the JWT signature, the JWT's `request_body_sha256` payload claim is compared against your own calculated sha256 hash of the event body
+
+=== Example Webhook Listener Code
+
+The following code demonstrates webhook signature verification with a simple Node server.
+
+[source,javascript]
+.Example Node.js Webhook Signature Verifier
+----
+const bodyParser = require('body-parser');
```

Comment:
Sometimes for larger examples we put the JavaScript in an example repository and then include it from the raw GitHub URL like some of the WebAuthn stuff:
```
https://raw.githubusercontent.com/FusionAuth/fusionauth-example-javascript-webauthn/main/buffer-to-base64url.js
```

Not sure whether we want to do that with this one or not. I see that there is other JS for writing a webhook that is pretty lengthy.

Reply:
good question.

Reply:
I would love to have this code in a separate repo, maybe even with tests?

Reply:
Done!  (without the tests)